### ISSUE TITLE
Model.fit() + LearningRateSchedule ignores initial_epoch/steps_per_epoch, giving wrong learning rates when resuming from checkpoints

### ISSUE THREAD

**Issue by chrisbutner** [tensorflow/tensorflow#49208]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes (see below)
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Google Colab, Ubuntu 20.04, Windows 10
- TensorFlow installed from (source or binary): (preinstalled), (custom package), binary (pip)
- TensorFlow version (use command below): v2.4.1-0-g85c8b2a817f 2.4.1, unknown 2.6.0, v2.4.0-49-g85c8b2a817f 2.4.1
- Python version: Python 3.7.10, Python 3.8.5, Python 3.7.9

**Describe the current behavior**

When training using a LearningRateSchedule, Optimizer and Model.fit(), the learning rate schedule ignores "initial_epoch" and "steps_per_epoch", so when starting a new session to resume training from a checkpoint, OptimizerV2._iterations and the resulting learning rate will be incorrect.

**Describe the expected behavior**

Model.fit() sees that "initial_epoch" and "steps_per_epoch" have been specified, and assigns "initial_epoch * steps_per_epoch" to OptimizerV2._iterations so that training can resume correctly.

**Standalone code to reproduce the issue**

```
import tensorflow as tf

input = tf.keras.layers.Input(shape=(2), dtype=tf.float32)
output = tf.keras.layers.Dense(1, activation="relu")(input)
model = tf.keras.Model(input, output)

learning_rate_schedule = tf.keras.optimizers.schedules.PiecewiseConstantDecay([10], [0.5, 0.25])
optimizer = tf.keras.optimizers.SGD(learning_rate=learning_rate_schedule)
model.compile(optimizer=optimizer, loss="mse")

class ValidationCallback(tf.keras.callbacks.Callback):

  def __init__(self, expected):
    self.expected = expected
    self.actual = []

  def on_train_batch_end(self, batch, logs=None):
    self.actual.append(tf.keras.backend.get_value(self.model.optimizer.learning_rate(self.model.optimizer._iterations)).item())

  def on_train_end(self, logs=None):
    if self.actual == self.expected:
      print("Good")
    else:
      print(f"Bug: Expected {self.expected}, actual {self.actual}")

values = tf.range(20, dtype=tf.float32)
inputs = tf.stack([values, values], axis=1)
outputs = values * 2.0
dataset = tf.data.Dataset.from_tensor_slices((inputs, outputs))
dataset = dataset.batch(1)

# Call fit() for steps 1-20.
print("Test 1")
model.fit(dataset, epochs=2, initial_epoch=0, steps_per_epoch=10, callbacks=[ValidationCallback([0.5] * 10 + [0.25] * 10)], verbose=0)

# Call fit() for steps 1-20 again (bug: the optimizer just retains its iterations, rather than calculating from epochs + initial_epoch + steps_per_epoch).
print("Test 2")
model.fit(dataset, epochs=2, initial_epoch=0, steps_per_epoch=10, callbacks=[ValidationCallback([0.5] * 10 + [0.25] * 10)], verbose=0)

# Call fit() for steps 1-20 a third time (work around the bug by resetting iterations).
print("Test 3")
model.optimizer.iterations.assign(0, read_value=False)
model.fit(dataset, epochs=2, initial_epoch=0, steps_per_epoch=10, callbacks=[ValidationCallback([0.5] * 10 + [0.25] * 10)], verbose=0)

# Call fit() for steps 11-20 (bug: should calculate iterations from epochs + initial_epoch + steps_per_epoch - very important when resuming training from a checkpoint).
print("Test 4")
model.optimizer.iterations.assign(0, read_value=False)
model.fit(dataset, epochs=2, initial_epoch=1, steps_per_epoch=10, callbacks=[ValidationCallback([0.25] * 10)], verbose=0)

# Call fit() for steps 11-20 (work around the bug by manually initializing iterations)
print("Test 5")
initial_epoch = 1
steps_per_epoch = 10
model.optimizer.iterations.assign(initial_epoch * steps_per_epoch, read_value=False)
model.fit(dataset, epochs=2, initial_epoch=initial_epoch, steps_per_epoch=steps_per_epoch, callbacks=[ValidationCallback([0.25] * 10)], verbose=0)
```

**Other info / logs**

```
Test 1
Good
Test 2
Bug: Expected [0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.25, 0.25, 0.25, 0.25, 0.25, 0.25, 0.25, 0.25, 0.25, 0.25], actual [0.25, 0.25, 0.25, 0.25, 0.25, 0.25, 0.25, 0.25, 0.25, 0.25, 0.25, 
0.25, 0.25, 0.25, 0.25, 0.25, 0.25, 0.25, 0.25, 0.25]
Test 3
Good
Test 4
Bug: Expected [0.25, 0.25, 0.25, 0.25, 0.25, 0.25, 0.25, 0.25, 0.25, 0.25], actual [0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5]
Test 5
Good
```

**Comment by saikumarchalla** [comment]
Was able to reproduce the issue using TF 2.4 and Nighty version as well. Please find the gist [here](https://colab.research.google.com/gist/saikumarchalla/d15d443885f2d85552b14cc5c9d6f594/untitled72.ipynb).Thanks!

**Comment by bhack** [comment]
I think that partially could be connected with https://github.com/tensorflow/tensorflow/issues/48947 as iterations probably is not a problem with a saved and resumed model. 
So I don't know if we will have other side effects manipulating `initial_epoch` on the same object. 
@chrisbutner Can you reproduce exactly the same training session (loss/acc) with your `model.optimizer.iterations.assign(0, read_value=False)` workaround?

**Comment by chrisbutner** [comment]
I'm not sure what you mean by "reproduce exactly the same training session", sorry. Thanks for that link - it sounds like https://github.com/tensorflow/tensorflow/issues/48947 deals with a very similar issue - that one is worse because it's the way a 10-line tutorial would save and resume training progress, and it's very clear what should happen - "iterations" should be saved and loaded.

In my case, I have a model that I can't save and load to a SavedModel format, so whenever I need to train or use it, I construct it from code and use `save_weights` and `load_weights` to save/resume training progress.

The general problem seems to be that a `LearningRateSchedule` needs to be told by the model/optimizer exactly how "old" the model is in steps and/or what the "absolute" step number is for each batch, and the API doesn't provide a clear way to do that. If you pass a dataset to `Model.fit` with 1000 examples, are those examples 1 to 1000, or 500,001 to 501,000? That information is crucial, and the current setup doesn't make it clear which learning rate the schedule will decide to use, at API doc time or runtime, which could result in a lot of wasted work for people.

If I were to address this problem, I would look at different ways of answering the question, "which absolute step numbers are these?", using prioritized signal/information.
- At the highest priority, passing `initial_epoch` and `steps_per_epoch` to `Model.fit` tells the model exactly which absolute step numbers are involved, even if the model along with its optimizer was saved and then loaded. That is, the `iterations` derived from `initial_epoch` and `steps_per_epoch` would override `iterations` saved as part of the optimizer state. I'm not suggesting that people should pass these two parameters as the primary way to signal absolute step numbers, but when they are passed, they provide clear signal.
- At the second level of priority, if a model is saved with optimizer state included, that can be used to resume from the correct absolute step number, and this should be trusted even when `epochs=N` is passed in, as long as `initial_epoch` is left as the default, if default and zero can be differentiated (i.e. treat the default 0->N as saved->saved + N).
- Finally, if `Model.fit` is just given a bunch of examples without any other clear signal, the optimizer has to assume that this is a fresh model starting from step 1. That may be a reasonable guess, but it would be better to have a clear way in the API to set absolute step number when resuming, rather than relying on `OptimizerV2.iterations` as an implementation detail. This is especially relevant to `Model.train_on_batch`.

**Comment by bhack** [comment]
> I'm not sure what you mean by "reproduce exactly the same training session", sorry.

I meant if the optimizer is exactly in the same status so if the loss/accuracy it is exactly the same at every step/epoch with your `iterations` reset workaround or we have others side effects. 
E.g. when serializing I still think that we have the old problem (see https://github.com/tensorflow/tensorflow/pull/42846) about recovering the optimizer status. So I don't know if we have other side effect with the `initial_epoch` if it was designed/tested only when re-loading saved models.

**Comment by chrisbutner** [comment]
Got it. Yes, you can definitely lose state; e.g., when I construct the model/optimizer/schedule and use `load_weights` + `initial_epoch` + `steps_per_epoch` + `OptimizerV2.iterations` workaround, I'm losing SGD momentum state and just live with that, because the wrong learning rate is disastrous, but uninitialized momentum should fix itself up quickly, and I'm not giving TensorFlow any info on how to reconstruct it.

In terms of API contract and expectations, I think TensorFlow should do everything it can *with the information it has* to follow instructions and maintain trust. In this case I think it's reasonable to expect the correct learning rate, but unreasonable to expect perfectly deterministic outputs.

If you save a model and optimizer, then all of the model and optimizer state should be loaded, or the API should make it clear what wouldn't be restored.

If you reconstruct the model from code and use `load_weights` + `initial_epoch` + `steps_per_epoch`, then TensorFlow doesn't have the information necessary to reconstruct momentum state, but it does have the information necessary to know that the user intends to resume from 100k steps and wants a learning rate of 0.02 and not 0.2.

This doesn't necessarily mean that a bunch of special cases should be set up to fix one-offs, but instead that care should go into choosing inputs and what they mean. A learning rate schedule needs an absolute step, not a per-session step, so where do we get that? Does `OptimizerV2.iterations` reflect that concept, or do we need to reconstruct it in one central place as best we can based on everything that went into building/compiling/restoring/fitting the model?

I'm not sure I understand your point about `initial_epoch` and side-effects, but if there are concerns about trusting existing `Model.fit` parameters when deciding on an absolute starting step, it may mean that TensorFlow needs more information; e.g. via a new parameter to the `LearningRateSchedule` constructor, optimizer, or `Model.fit`. Silently guessing and getting it wrong is very costly for users in the product space that TensorFlow operates in.

**Comment by bhack** [comment]
What I meant is all these Is related to the optimizer lifecycle.
If you don't care about the status/reproducibilty of the optimizer iterations is related to the lifecycle of the object and not to the model or fit.
Yes probably a warning could be printed if this is the intended behaviour when you have `initial_,epoch` but with your multiple fit  example and without serialization probably It expect a new optimizer. 
E.g. https://stackoverflow.com/questions/56806419/keras-how-to-reset-optimizer-state

**Comment by chrisbutner** [comment]
Sure, I think that lines up with what I'm saying - maybe I need to explicitly tell the optimizer which absolute step to start with when starting a new `Model.fit` (just to be clear, I'm not doing multiple fit()s, that's just in the repro example).

The problem is that (a) it's not documented that `LearningRateSchedule` relies on `Optimizer.iterations`, and (b) `iterations` isn't on the public API for, e.g., https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/SGD

So, I'm happy to manually initialize `Optimizer.iterations`, but I have no way of knowing whether e.g. TF2.7 would break my code.

**Comment by bhack** [comment]
As iterations is in the weights and in you real case you load the weights I think that you need to have a fresh new optimizer and re-compile after weights load

**Comment by chrisbutner** [comment]
Ah, thank you, I didn't realize `save_weights`/`load_weights` also saved optimizer state: I thought that was restricted to `Model.save`/`models.load_model`, but I see it documented there.

After adjusting the repro, everything seems to work as expected regardless of the order of `compile` and `load_weights`.

The reason I couldn't automatically take advantage of the optimizer weights is because I was saving/loading weights using the full model but using a subset of it for compiling and training, wrapped with a new `Model`, without realizing the consequences.

I think TF is behaving as expected then and this issue can be closed (it's confusing then that https://github.com/tensorflow/tensorflow/issues/48947 has a problem with iterations via `Model.save`/`models.load_model`). Thank you for explaining.

**Comment by sachinprasadhs** [comment]
@chrisbutner , Closing the issue since it is working as expected and your issue is resolved, Thanks!
